classes/class4/README.md

```diff
@@ -1 +1,484 @@
+by refactoring it a bit we were able to reuse most of the same code for a second purpose.
+
+Secondly, we added a check in the `getCatGET` function to determine if the request
+was an XHR request and if so, to return the relative path to the image, instead of the full image file.
+
+Now we need to write some code on the client in order to get the 3rd form to send the XHR request.
+
+Replace the contents of `main.js` with this:
+```javascript
+var $form = $("#ajax-form");
+
+var onSuccess = function(data, status) {
+  var img = "<img src='"+data+"'/>";
+  $("#result").html(img);
+};
+
+var onError = function(data, status) {
+  console.log("status", status);
+  console.log("error", data);
+};
+
+$form.submit(function(event) {
+  event.preventDefault();
+  var mood = $form.find("[name='mood']:checked").val();
+  var name = $form.find("[name='name']").val();
+  $.get("getCat", {
+    mood: mood,
+    name: name
+  })
+    .done(onSuccess)
+    .error(onError);
+});
+```
+
+Let's work through this file one step at a time:
+* First we store the jQuery selector for the 3rd form so we don't need to keep searching for it.
+* We define an `onSuccess` function. this takes `data` and `status`, creates a new `<img>` tag as a string
+and the sets the contents the `<div id=result>` element to that image.
+* We define an `onError` function, which prints the status and contents of the error.
+* We add an event handler for the "submit" event to the 3rd form.
+This means that if that form raises a "submit" event then the callback function will be run with the event object as a parameter.
+  * The first thing we do in the callback is `event.preventDefault();`.
+  This line stops the default action of the event from being triggered.
+  In this case, this is what stops the browser from sending the request and loading a new page.
+  * Using some jQuery DOM searching we get the values of the "name" text input and the "mood" radio button.
+  * We then initiate an AJAX request using the GET method with `$.get()`. jQuery similarly has a `$.post()` function for POST.
+  We pass the two parameters into the GET request which are internally packaged into the same query string we've seen before.
+  * Lastly, we specify what should happen if the request is successful or if it returns with an error.
+  The `$.get()` function returns a jQuery XHR object ('jqXHR') which implements a Promise interface.
+  Promises are a structure that allow management of lots of callbacks in a simple, easier to read manner.
+  We wont talk too much about promises for now, but the important part is to know that the `.done()` callback is called
+  when the jqXHR object is completed successfully, the `.error()` callback is called
+  when there is an error, and the `.always()` callback (which we did not include here)
+  is called on completion whether successful or not.
+
+  All of this put together allows us to communicate with the server without sending browser requests,
```

Comment:
without sending browser requests -> without sending `GET` requests from the URL bar, and without reloading the page